python/ray/air/preprocessor.py

```diff
@@ -165,18 +166,62 @@ def _fit(self, dataset: Dataset) -> "Preprocessor":
     def _transform(self, dataset: Dataset) -> Dataset:
         # TODO(matt): Expose `batch_size` or similar configurability.
         # The default may be too small for some datasets and too large for others.
-        return dataset.map_batches(self._transform_pandas, batch_format="pandas")
+        dataset_format = dataset._dataset_format()
+        exc_msg = "Neither `_transform_arrow` nor `_transform_pandas` are implemented."
+
+        # If only _transform_arrow is implemented, will convert the data to arrow.
+        # If only _transform_pandas is implemented, will convert the data to pandas.
+        # If both are implemented, will pick the method corresponding to the format
+        # for best performance.
+        # Implementation is defined as overriding the method in a sub-class.
+        if dataset_format == "arrow":
+            if self.__class__._transform_arrow != Preprocessor._transform_arrow:
+                return dataset.map_batches(
+                    self._transform_arrow, batch_format="pyarrow"
+                )
+            elif self.__class__._transform_pandas != Preprocessor._transform_pandas:
+                return dataset.map_batches(
+                    self._transform_pandas, batch_format="pandas"
+                )
+            else:
+                raise NotImplementedError(exc_msg)
+        elif dataset_format == "pandas":
+            if self.__class__._transform_pandas != Preprocessor._transform_pandas:
+                return dataset.map_batches(
+                    self._transform_pandas, batch_format="pandas"
+                )
+            elif self.__class__._transform_arrow != Preprocessor._transform_arrow:
+                return dataset.map_batches(
+                    self._transform_arrow, batch_format="pyarrow"
+                )
+            else:
+                raise NotImplementedError(exc_msg)
+        else:
+            raise ValueError(
+                f"Unsupported Dataset format: '{dataset_format}'. Only 'pandas' and "
+                "'arrow' Dataset formats are supported."
+            )
 
     def _transform_batch(self, df: DataBatchType) -> DataBatchType:
         import pandas as pd
 
-        # TODO(matt): Add `_transform_arrow` to use based on input type.
-        # Reduce conversion cost if input is in Arrow:  Arrow -> Pandas -> Arrow.
-        if not isinstance(df, pd.DataFrame):
+        try:
+            import pyarrow
+        except ImportError:
+            pyarrow = None
+
+        if isinstance(df, pd.DataFrame):
+            return self._transform_pandas(df)
+        elif pyarrow is not None and isinstance(df, pyarrow.Table):
+            return self._transform_arrow(df)
```

Comment:
If it's not implemented, then it will just throw a NotImplementedError when it goes into it. For `transform_batch`, I feel like the user should convert the data themselves. I can make this exception more informative, though.

Reply:
I'm worried that UX might not be ideal, especially if the user doesn't have control of the batch before it's passed into `_transform_batch()`. When exactly will `_transform_batch()` be used? I did a quick repo search and `Preprocessor._transform_batch()` is only used by `Chain._transform_batch()`, which doesn't appear to be used by anything else.

Reply:
+1 on doing the method check, and if `_transform_arrow` is not implemented, then converting to pandas. (you can use the conversion utils here https://github.com/ray-project/ray/pull/25534)

As a user I don't really care whether _transform_arrow is implemented or _transform_pandas is implemented. I just want to call transform_batch and expect it to work.

Reply:
@clarkzinzow `_transform_batch` is used by `transform_batch`. It's intended as a way to directly transform data batches that are not a part of a Ray Dataset

Reply:
Ok, that makes sense. Let me do that

Reply:
I think that you shouldn't convert back to the input type, in order to minimize conversions in the case of chained preprocessors.